### DISCUSSION TITLE
Exclude members in nested list by string path

### DISCUSSION THREAD

**Discussion by q34hfq34fq5374rqhdf0vih3435dd23234wkklo** [fluentassertions/fluentassertions#2671] [Q&A]
Hi,

**In this example:** https://fluentassertions.com/objectgraphs/#selecting-members you can exclude by path which works for my non-collection members.

```cs
orderDto.Should().BeEquivalentTo(order, options => options 
    .Excluding(ctx => ctx.Path == "Level.Level.Text"));
```

**How would you exclude the Employee.Role.id in this class:**
```cs
  public class Employee
  {
    public int id { get; set; }
    public string name { get; set; }
    public List<Role> roles { get; set; }
  }

  public class Role
  {
    public int id { get; set; }
    public string name { get; set; }
  }
```

**I have tried:**
```cs
        firstSource.Should().BeEquivalentTo(secondSource, options => options
    .Excluding(ctx => ctx.Path == "Employee.roles.id"));

        firstSource.Should().BeEquivalentTo(secondSource, options => options
    .Excluding(ctx => ctx.Path == "roles.id"));

        firstSource.Should().BeEquivalentTo(secondSource, options => options
    .Excluding(ctx => ctx.Path == "firstSource.roles.id"));
```

**I still get a message like (meaning the propbery was not excluded):**
```
Expected property firstSource.roles[0].id to be 123456, but found <null>.
```

I am able to get this to work with the some other exclude methods, however I would like to use string-based paths because another library I am using (https://github.com/replaysMike/AnyDiff) can take a collection of strings and it would be nice to be able to use the same collection for both libraries.

Thank you.

**Comment by ITaluone** [comment]
Why don't you use the fluent API (see https://github.com/fluentassertions/fluentassertions/pull/1782)?

```csharp
[Fact]
public void Test()
{
    var e = new Employee();

    e.Should().BeEquivalentTo(new Employee(), opt => opt
        .For(employee => employee.roles)
        .Exclude(role => role.id));
}
```

**Reply by q34hfq34fq5374rqhdf0vih3435dd23234wkklo** [reply]
Hi,

The fluent APIs are nice and I have been using them. I was able to achieve the desired behavior with the fluent ones, however I would like to use the string-based ones because I could use the same set of strings for both libraries (this and AnyDiff).

I have a complex nested object with many collections and making the exclude/include logic is quite long so using one source would be nice.

Thanks

**Reply by IT-VBFK** [reply]
Hmm..

can you try:

```csharp
[Fact]
public void Test()
{
    var e = new Employee();

    e.Should().BeEquivalentTo(new Employee(), opt => opt
    .Excluding(ctx => ctx.Path == "Employee.roles[].id"));
}
```

**Reply by q34hfq34fq5374rqhdf0vih3435dd23234wkklo** [reply]
Tried that with a few variants and wasn't able to get it to work. I will continue to play around with this.

Thank you

**Reply by dennisdoomen** [reply]
I couldn't find a good example so quickly, but I'm pretty sure you can use `.Excluding(ctx => ctx.Path == "roles[].id"));`

**Reply by q34hfq34fq5374rqhdf0vih3435dd23234wkklo** [reply]
That didn't work for me. If I find a working version I will post it. Thanks

**Comment by dennisdoomen** [comment]
You can use what FA [internally uses](https://github.com/fluentassertions/fluentassertions/blob/180c1a33aa41538674fad623c37af4bb20c586ab/Src/FluentAssertions/Common/StringExtensions.cs#L42) for solving this with expression trees, 

firstSource.Should().BeEquivalentTo(secondSource, options => options
    .Excluding(ctx => Regex.Replace(ctx.Path, @"\[[0-9]+\]", "[]") == "Roles[].Id"));
